### PR TITLE
Optimise branch mapper 

### LINKED ISSUE #3450: Optimise branch mapper
- allocate less memory, and early out

### PR DESCRIPTION
**Pull request by ugras-ergun-sonarsource**
fixes #3450

### CODE REVIEW

File: src/ConnectedMode/ServerBranchProvider.cs [hunk 1]
```diff
@@ -81,7 +81,7 @@ public async Task<string> GetServerBranchNameAsync(CancellationToken token)
             }
 
             var repo = createRepo(gitRepoRoot);
-            branchName = await branchMatcher.GetMatchingBranch(config.Project.ProjectKey, repo);
+            branchName = await branchMatcher.GetMatchingBranch(config.Project.ProjectKey, repo, token);
```
Comment (ugras-ergun-sonarsource) [inline src/ConnectedMode/ServerBranchProvider.cs:84]: for some reason token was missing after the merge

File: src/ConnectedMode.UnitTests/ServerBranchProviderTests.cs [hunk 2]
```diff
@@ -118,7 +118,7 @@ public async Task Get_ConnectedModeAndHasGitRepo_HasMatchingBranch_ReturnsExpect
             configProvider.VerifyAll();
             gitWorkspace.Verify(x => x.GetRepoRoot(), Times.Once);
             createRepoOp.Verify(x => x.Invoke("c:\\aaa\\reporoot"), Times.Once);
-            branchMatcher.Verify(x => x.GetMatchingBranch("my project key", repo), Times.Once);
+            branchMatcher.Verify(x => x.GetMatchingBranch("my project key", repo, It.IsAny<CancellationToken>()), Times.Once);
```
Comment (ugras-ergun-sonarsource) [inline src/ConnectedMode.UnitTests/ServerBranchProviderTests.cs:121]: for some reason token was missing after the merge

File: src/ConnectedMode.UnitTests/ServerBranchProviderTests.cs [hunk 3]
```diff
@@ -146,7 +146,7 @@ public async Task Get_ConnectedModeAndHasGitRepo_NoMatchingBranch_ReturnsNull()
             var actual = await testSubject.GetServerBranchNameAsync(CancellationToken.None);
 
             actual.Should().BeNull();
-            branchMatcher.Verify(x => x.GetMatchingBranch("my project key", repo), Times.Once);
+            branchMatcher.Verify(x => x.GetMatchingBranch("my project key", repo, It.IsAny<CancellationToken>()), Times.Once);
```
Comment (ugras-ergun-sonarsource) [inline src/ConnectedMode.UnitTests/ServerBranchProviderTests.cs:149]: for some reason token was missing after the merge

File: src/ConnectedMode.UnitTests/ServerBranchProviderTests.cs [hunk 4]
```diff
@@ -196,7 +196,7 @@ private static Mock<ServerBranchProvider.CreateRepositoryObject> CreateCreateRep
         private static Mock<IBranchMatcher> CreateBranchMatcher(string branchToReturn)
         {
             var branchMatcher = new Mock<IBranchMatcher>();
-            branchMatcher.Setup(x => x.GetMatchingBranch(It.IsAny<string>(), It.IsAny<IRepository>()))
+            branchMatcher.Setup(x => x.GetMatchingBranch(It.IsAny<string>(), It.IsAny<IRepository>(), It.IsAny<CancellationToken>()))
```
Comment (ugras-ergun-sonarsource) [inline src/ConnectedMode.UnitTests/ServerBranchProviderTests.cs:199]: for some reason token was missing after the merge

**Review by duncanp-sonar — APPROVED** [review]

**Pull request merged**